src/octoprint/util/comm.py

```diff
@@ -3126,134 +3184,213 @@ def _enqueue_for_sending(self, command, linenumber=None, command_type=None, on_s
+						# at this point if our core code contains a bug
+						assert len(results) == 1
 
-							# tickle...
-							self._continue_sending()
+						# we only use the first (and only!) entry here
+						command, _, gcode, subcode, tags = results[0]
 
-							# ... and fetch the next item
-							continue
+					if command.strip() == "":
+						self._logger.info("Refusing to send an empty line to the printer")
 
-						# now comes the part where we increase line numbers and send stuff - no turning back now
-						self._do_send(command, gcode=gcode)
+						# same here, tickle the queues manually
+						self._continue_sending()
 
-					# trigger "sent" phase and use up one "ok"
-					if on_sent is not None and callable(on_sent):
-						# we have a sent callback for this specific command, let's execute it now
-						on_sent()
-					self._process_command_phase("sent", command, command_type, gcode=gcode, subcode=subcode, tags=tags)
+						# and fetch the next item
+						return True
 
-					# we only need to use up a clear if the command we just sent was either a gcode command or if we also
-					# require ack's for unknown commands
-					use_up_clear = self._unknownCommandsNeedAck
-					if gcode is not None:
-						use_up_clear = True
+					# handle @ commands
+					if gcode is None and command.startswith("@"):
+						self._process_atcommand_phase("sending", command, tags=tags)
 
-					if use_up_clear:
-						# if we need to use up a clear, do that now
-						self._clear_to_send.clear()
-					else:
-						# Otherwise we need to tickle the read queue - there might not be a reply
-						# to this command, so our _monitor loop will stay waiting until timeout. We
-						# definitely do not want that, so we tickle the queue manually here
+						# tickle...
 						self._continue_sending()
 
-				finally:
-					# no matter _how_ we exit this block, we signal that we
-					# are done processing the last fetched queue entry
-					self._send_queue.task_done()
+						# ... and fetch the next item
+						return True
+
+					# now comes the part where we increase line numbers and send stuff - no turning back now
+					self._do_send(command, gcode=gcode)
+
+				# trigger "sent" phase and use up one "ok"
+				if on_sent is not None and callable(on_sent):
+					# we have a sent callback for this specific command, let's execute it now
+					on_sent()
+				self._process_command_phase("sent", command, command_type, gcode=gcode, subcode=subcode, tags=tags)
+
+				# we only need to use up a clear if the command we just sent was either a gcode command or if we also
+				# require ack's for unknown commands
+				use_up_clear = self._unknownCommandsNeedAck
+				if gcode is not None:
+					use_up_clear = True
+
+				if use_up_clear:
+					# if we need to use up a clear, do that now
+					self._clear_to_send.clear()
+
+					if ( self._advanced_ok_detected and self._state == self.STATE_PRINTING):
+						with self._line_mutex:
+							linenumber = self._current_line - 1
```

Comment:
You'll have to take into account that a line might have been sent without a number (possible for in-band commands that aren't GCODE, or just for certain configuration options).